types/wlr_pointer_constraints_v1.c

```diff
@@ -133,58 +159,42 @@ static const struct zwp_locked_pointer_v1_interface locked_pointer_impl = {
 	.set_region               = pointer_constraint_set_region,
 };
 
-static void pointer_constraints_lock_pointer(struct wl_client *client,
-		struct wl_resource *cons_resource,
-		uint32_t id,
-		struct wl_resource *surface,
-		struct wl_resource *pointer,
-		struct wl_resource *region,
-		uint32_t lifetime) {
-	struct wlr_pointer_constraints_v1 *cons = pointer_constraints_from_resource(cons_resource);
+static void pointer_constraint_create(struct wl_client *client,
+		struct wl_resource *pointer_constraints_resource, uint32_t id,
+		struct wl_resource *surface_resource, struct wl_resource *pointer,
+		struct wl_resource *region, uint32_t lifetime,
+		enum wlr_pointer_constraint_v1_type type) {
+	struct wlr_pointer_constraints_v1 *pointer_constraints = pointer_constraints_from_resource(pointer_constraints_resource);
+	struct wlr_surface *surface = wlr_surface_from_resource(surface_resource);
 
-	struct wlr_pointer_constraint_v1 *constraint =
-		calloc(1, sizeof(struct wlr_pointer_constraint_v1));
-	if (constraint == NULL) {
-		wl_client_post_no_memory(client);
-		return;
-	}
+	struct wl_list *list = &surface->pointer_constraints;
 
-	int version = wl_resource_get_version(cons_resource);
-	constraint->resource = wl_resource_create(client,
-		&zwp_locked_pointer_v1_interface, version, id);
-	constraint->surface         = surface;
-	constraint->pointer         = pointer;
-	constraint->current.region = region;
-	constraint->pending.region = region;
-	constraint->lifetime        = lifetime;
-	constraint->type            = WLR_POINTER_CONSTRAINT_V1_LOCKED;
+	int version = wl_resource_get_version(pointer_constraints_resource);
+
+	bool locked_pointer = type == WLR_POINTER_CONSTRAINT_V1_LOCKED;
 
-	wl_list_init(&constraint->link);
-	wl_signal_init(&constraint->events.destroy);
+	struct wl_resource *resource = locked_pointer ?
+		wl_resource_create(client, &zwp_locked_pointer_v1_interface, version, id) :
+		wl_resource_create(client, &zwp_confined_pointer_v1_interface, version, id);
 
-	if (constraint->resource == NULL) {
-		free(constraint);
+	if (resource == NULL) {
 		wl_client_post_no_memory(client);
 		return;
 	}
 
-	wl_resource_set_implementation(constraint->resource,
-		&locked_pointer_impl, constraint, pointer_constraint_destroy_resource);
-
-	wlr_log(L_DEBUG, "new locked_pointer %p (res %p)", constraint,
-		constraint->resource);
-
-	wlr_signal_emit_safe(&cons->events.constrain, constraint);
-}
-
-static void pointer_constraints_confine_pointer(struct wl_client *client,
-		struct wl_resource *cons_resource,
-		uint32_t id,
-		struct wl_resource *surface,
-		struct wl_resource *pointer,
-		struct wl_resource *region,
-		uint32_t lifetime) {
-	struct wlr_pointer_constraints_v1 *cons = pointer_constraints_from_resource(cons_resource);
+	struct wlr_seat_client *seat_client = wlr_seat_client_from_pointer_resource(pointer);
+
+	{
```

Comment:
This doesn't need its own block, just call it _constraint or reuse the variable or something.